File: android/app/src/main/java/com/mobile/ScreenTimeManager.kt
package com.mobile

import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.os.Handler
import android.os.Looper
import androidx.core.content.ContextCompat
import org.json.JSONArray
import org.json.JSONObject
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone
import java.util.UUID
import kotlin.math.max
import kotlin.math.min

data class ScreenTimeGrant(
  val grantId: String,
  val childUserId: String,
  val familyId: String,
  val grantedMinutes: Int,
  val grantedAt: Long,
  val expiresAt: Long,
  var status: String,
  var revokedAt: Long?,
  var allowedPackages: List<String>,
  var allowedCategories: List<String>,
  val grantedByUserId: String,
  val sourceTaskId: String?,
  var usedMinutesOverride: Int?
)

class ScreenTimeManager(private val context: Context) {
  private val prefs: SharedPreferences =
    context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
  private val handler = Handler(Looper.getMainLooper())
  private val expiryRunnable = Runnable { expireGrantsIfNeeded() }

  var onUpdate: ((ScreenTimeGrant) -> Unit)? = null

  fun grantScreenTime(
    childUserId: String,
    durationMinutes: Int,
    allowedPackages: List<String>?,
    allowedCategories: List<String>?
  ): ScreenTimeGrant {
    if (childUserId.isBlank()) {
      throw IllegalArgumentException("Child user ID is required.")
    }
    if (durationMinutes <= 0) {
      throw IllegalArgumentException("Duration must be greater than zero.")
    }

    val now = System.currentTimeMillis()
    val expiresAt = now + durationMinutes * 60L * 1000L
    val grant = ScreenTimeGrant(
      grantId = UUID.randomUUID().toString(),
      childUserId = childUserId,
      familyId = "",
      grantedMinutes = durationMinutes,
      grantedAt = now,
      expiresAt = expiresAt,
      status = "active",
      revokedAt = null,
      allowedPackages = allowedPackages ?: emptyList(),
      allowedCategories = allowedCategories ?: emptyList(),
      grantedByUserId = "local",
      sourceTaskId = null,
      usedMinutesOverride = null
    )

    val grants = loadGrants().toMutableList().apply {
      removeAll { it.childUserId == childUserId }
      add(grant)
    }
    saveGrants(grants)

    applyGrantSettings()
    scheduleExpiryTimer()
    onUpdate?.invoke(grant)
    return grant
  }

  fun revokeScreenTime(childUserId: String): ScreenTimeGrant? {
    if (childUserId.isBlank()) {
      throw IllegalArgumentException("Child user ID is required.")
    }

    val now = System.currentTimeMillis()
    var revokedGrant: ScreenTimeGrant? = null
    val grants = loadGrants().map { grant ->
      if (grant.childUserId == childUserId) {
        val updated = grant.copy(
          status = "revoked",
          revokedAt = now,
          usedMinutesOverride = usedMinutes(grant, now)
        )
        revokedGrant = updated
        updated
      } else {
        grant
      }
    }.filter { it.childUserId != childUserId }

    saveGrants(grants)
    applyGrantSettings()
    scheduleExpiryTimer()
    revokedGrant?.let { onUpdate?.invoke(it) }
    return revokedGrant
  }

  fun getScreenTimeStatus(childUserId: String?): Pair<Int, List<ScreenTimeGrant>> {
    expireGrantsIfNeeded()
    val now = System.currentTimeMillis()
    val activeGrants = loadGrants().filter { grant ->
      grant.status == "active" && grant.expiresAt > now &&
        (childUserId == null || grant.childUserId == childUserId)
    }

    val remaining = activeGrants
      .map { remainingMinutes(it, now) }
      .maxOrNull() ?: 0

    return Pair(remaining, activeGrants)
  }

  fun syncGrantFromServer(grantData: Map<String, Any?>): ScreenTimeGrant {
    val grantId = grantData["grantId"] as? String
      ?: throw IllegalArgumentException("grantId is required.")
    val childUserId = grantData["childUserId"] as? String
      ?: throw IllegalArgumentException("childUserId is required.")
    val familyId = grantData["familyId"] as? String
      ?: throw IllegalArgumentException("familyId is required.")
    val grantedMinutes = intValue(grantData["grantedMinutes"])
    if (grantedMinutes <= 0) {
      throw IllegalArgumentException("grantedMinutes must be greater than zero.")
    }

    val grantedAt = dateMillis(grantData["grantedAt"])
      ?: throw IllegalArgumentException("grantedAt is required.")
    val expiresAt = dateMillis(grantData["expiresAt"])
      ?: throw IllegalArgumentException("expiresAt is required.")

    val status = grantData["status"] as? String ?: "active"
    val revokedAt = dateMillis(grantData["revokedAt"])
    val allowedPackages = (grantData["allowedAppBundleIds"] as? List<*>)
      ?.mapNotNull { it as? String }
      ?: (grantData["allowedPackages"] as? List<*>)?.mapNotNull { it as? String }
      ?: emptyList()
    val allowedCategories = (grantData["allowedCategories"] as? List<*>)
      ?.mapNotNull { it as? String } ?: emptyList()
    val grantedByUserId = grantData["grantedByUserId"] as? String ?: "server"
    val sourceTaskId = grantData["sourceTaskId"] as? String
    val usedMinutesOverride = intValue(grantData["usedMinutes"])

    var grant = ScreenTimeGrant(
      grantId = grantId,
      childUserId = childUserId,
      familyId = familyId,
      grantedMinutes = grantedMinutes,
      grantedAt = grantedAt,
      expiresAt = expiresAt,
      status = status,
      revokedAt = revokedAt,
      allowedPackages = allowedPackages,
      allowedCategories = allowedCategories,
      grantedByUserId = grantedByUserId,
      sourceTaskId = sourceTaskId,
      usedMinutesOverride = if (usedMinutesOverride > 0) usedMinutesOverride else null
    )

    val now = System.currentTimeMillis()
    if (grant.status == "active" && grant.expiresAt <= now) {
      grant = grant.copy(
        status = "expired",
        usedMinutesOverride = usedMinutes(grant, now)
      )
    }

    val grants = loadGrants().toMutableList().apply {
      removeAll { it.grantId == grantId }
      add(grant)
    }
    saveGrants(grants)

    applyGrantSettings()
    scheduleExpiryTimer()
    onUpdate?.invoke(grant)
    return grant
  }

  fun grantToMap(grant: ScreenTimeGrant): Map<String, Any?> {
    val now = System.currentTimeMillis()
    val usedMinutes = usedMinutes(grant, now)
    val remainingMinutes = remainingMinutes(grant, now)
    val payload = mutableMapOf<String, Any?>(
      "grantId" to grant.grantId,
      "childUserId" to grant.childUserId,
      "familyId" to grant.familyId,
      "grantedMinutes" to grant.grantedMinutes,
      "usedMinutes" to usedMinutes,
      "remainingMinutes" to remainingMinutes,
      "status" to grant.status,
      "grantedAt" to isoString(grant.grantedAt),
      "expiresAt" to isoString(grant.expiresAt),
      "grantedByUserId" to grant.grantedByUserId
    )

    if (grant.revokedAt != null) {
      payload["revokedAt"] = isoString(grant.revokedAt!!)
    }
    if (grant.sourceTaskId != null) {
      payload["sourceTaskId"] = grant.sourceTaskId
    }
    if (grant.allowedPackages.isNotEmpty()) {
      payload["allowedAppBundleIds"] = grant.allowedPackages
    }
    if (grant.allowedCategories.isNotEmpty()) {
      payload["allowedCategories"] = grant.allowedCategories
    }
    return payload
  }

  private fun applyGrantSettings() {
    val now = System.currentTimeMillis()
    val activeGrants = loadGrants().filter { grant ->
      grant.status == "active" && grant.expiresAt > now
    }

    if (activeGrants.isEmpty()) {
      startBlockerService()
      return
    }

    val allowAll = activeGrants.any { grant ->
      grant.allowedPackages.isEmpty() && grant.allowedCategories.isEmpty() ||
        grant.allowedCategories.contains("all")
    }

    if (allowAll) {
      stopBlockerService()
      return
    }

    startBlockerService()
  }

  private fun startBlockerService() {
    val intent = Intent(context, AppBlockerService::class.java)
    ContextCompat.startForegroundService(context, intent)
  }

  private fun stopBlockerService() {
    context.stopService(Intent(context, AppBlockerService::class.java))
  }

  private fun scheduleExpiryTimer() {
    handler.removeCallbacks(expiryRunnable)
    val now = System.currentTimeMillis()
    val nextExpiry = loadGrants()
      .filter { it.status == "active" }
      .map { it.expiresAt }
      .minOrNull()
    if (nextExpiry != null) {
      val delay = max(1000L, nextExpiry - now)
      handler.postDelayed(expiryRunnable, delay)
    }
  }

  private fun expireGrantsIfNeeded() {
    val now = System.currentTimeMillis()
    val expired = mutableListOf<ScreenTimeGrant>()
    val grants = loadGrants().map { grant ->
      if (grant.status == "active" && grant.expiresAt <= now) {
        val updated = grant.copy(
          status = "expired",
          usedMinutesOverride = usedMinutes(grant, now)
        )
        expired.add(updated)
        updated
      } else {
        grant
      }
    }
    if (expired.isNotEmpty()) {
      saveGrants(grants)
      applyGrantSettings()
      expired.forEach { onUpdate?.invoke(it) }
    }
    scheduleExpiryTimer()
  }

  private fun usedMinutes(grant: ScreenTimeGrant, now: Long): Int {
    grant.usedMinutesOverride?.let {
      return min(grant.grantedMinutes, max(0, it))
    }
    val elapsedMinutes = ((now - grant.grantedAt) / 60000L).toInt()
    return min(grant.grantedMinutes, max(0, elapsedMinutes))
  }

  private fun remainingMinutes(grant: ScreenTimeGrant, now: Long): Int {
    if (grant.status != "active" || now >= grant.expiresAt) {
      return 0
    }
    return max(0, grant.grantedMinutes - usedMinutes(grant, now))
  }

  private fun loadGrants(): List<ScreenTimeGrant> = loadGrantsFromPrefs(prefs)

  private fun saveGrants(grants: List<ScreenTimeGrant>) {
    val array = JSONArray()
    grants.forEach { grant ->
      array.put(grantToJson(grant))
    }
    prefs.edit().putString(PREFS_KEY, array.toString()).apply()
  }

  private fun grantToJson(grant: ScreenTimeGrant): JSONObject {
    val obj = JSONObject()
    obj.put("grantId", grant.grantId)
    obj.put("childUserId", grant.childUserId)
    obj.put("familyId", grant.familyId)
    obj.put("grantedMinutes", grant.grantedMinutes)
    obj.put("grantedAt", grant.grantedAt)
    obj.put("expiresAt", grant.expiresAt)
    obj.put("status", grant.status)
    if (grant.revokedAt != null) {
      obj.put("revokedAt", grant.revokedAt)
    }
    obj.put("allowedPackages", JSONArray(grant.allowedPackages))
    obj.put("allowedCategories", JSONArray(grant.allowedCategories))
    obj.put("grantedByUserId", grant.grantedByUserId)
    if (grant.sourceTaskId != null) {
      obj.put("sourceTaskId", grant.sourceTaskId)
    }
    if (grant.usedMinutesOverride != null) {
      obj.put("usedMinutesOverride", grant.usedMinutesOverride)
    }
    return obj
  }

  private fun intValue(raw: Any?): Int {
    return when (raw) {
      is Number -> raw.toInt()
      is String -> raw.toIntOrNull() ?: 0
      else -> 0
    }
  }

  private fun dateMillis(raw: Any?): Long? {
    return when (raw) {
      is Number -> raw.toLong()
      is String -> parseIsoMillis(raw)
      is Date -> raw.time
      else -> null
    }
  }

  private fun isoString(timestamp: Long): String {
    return ISO_FORMAT.format(Date(timestamp))
  }

  private fun parseIsoMillis(value: String): Long? {
    return try {
      ISO_FORMAT.parse(value)?.time
    } catch (_: ParseException) {
      null
    }
  }

  companion object {
    private const val PREFS_NAME = "famconomy_screen_time"
    private const val PREFS_KEY = "famconomy_screen_time_grants_v1"
    private val ISO_FORMAT = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US).apply {
      timeZone = TimeZone.getTimeZone("UTC")
    }

    fun loadGrantsFromPrefs(context: Context): List<ScreenTimeGrant> {
      val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
      return loadGrantsFromPrefs(prefs)
    }

    private fun loadGrantsFromPrefs(prefs: SharedPreferences): List<ScreenTimeGrant> {
      val data = prefs.getString(PREFS_KEY, "[]") ?: "[]"
      val array = JSONArray(data)
      val grants = mutableListOf<ScreenTimeGrant>()
      for (i in 0 until array.length()) {
        val obj = array.optJSONObject(i) ?: continue
        grants.add(
          ScreenTimeGrant(
            grantId = obj.optString("grantId"),
            childUserId = obj.optString("childUserId"),
            familyId = obj.optString("familyId"),
            grantedMinutes = obj.optInt("grantedMinutes"),
            grantedAt = obj.optLong("grantedAt"),
            expiresAt = obj.optLong("expiresAt"),
            status = obj.optString("status", "active"),
            revokedAt = if (obj.has("revokedAt")) obj.optLong("revokedAt") else null,
            allowedPackages = jsonArrayToList(obj.optJSONArray("allowedPackages")),
            allowedCategories = jsonArrayToList(obj.optJSONArray("allowedCategories")),
            grantedByUserId = obj.optString("grantedByUserId", "server"),
            sourceTaskId = obj.optString("sourceTaskId", null),
            usedMinutesOverride = if (obj.has("usedMinutesOverride")) obj.optInt("usedMinutesOverride") else null
          )
        )
      }
      return grants
    }

    private fun jsonArrayToList(array: JSONArray?): List<String> {
      if (array == null) return emptyList()
      val list = mutableListOf<String>()
      for (i in 0 until array.length()) {
        list.add(array.optString(i))
      }
      return list
    }
  }
}

File: ios/Mobile/ScreenTimeManager.swift
import Foundation
import FamilyControls
import ManagedSettings

@available(iOS 16.0, *)
final class ScreenTimeManager {
  struct ScreenTimeGrantRecord: Codable {
    let grantId: String
    let childUserId: String
    let familyId: String
    let grantedMinutes: Int
    let grantedAt: Date
    let expiresAt: Date
    var status: String
    var revokedAt: Date?
    var allowedBundleIds: [String]
    var allowedCategories: [String]
    let grantedByUserId: String
    let sourceTaskId: String?
    var usedMinutesOverride: Int?
  }

  enum ScreenTimeError: LocalizedError {
    case invalidPayload
    case invalidDuration
    case missingChildUserId

    var errorDescription: String? {
      switch self {
      case .invalidPayload:
        return "Invalid payload."
      case .invalidDuration:
        return "Duration must be greater than zero."
      case .missingChildUserId:
        return "Child user ID is required."
      }
    }
  }

  private let settingsStore = ManagedSettingsStore()
  private let queue = DispatchQueue(label: "com.famconomy.familycontrols.grants")
  private let timerQueue = DispatchQueue(label: "com.famconomy.familycontrols.timer")
  private let defaultsKey = "famconomy_screen_time_grants_v1"
  private let dateFormatter = ISO8601DateFormatter()
  private var expiryTimer: DispatchSourceTimer?

  var onUpdate: ((ScreenTimeGrantRecord) -> Void)?

  init() {
    pruneExpiredGrants()
  }

  func requestAuthorization(completion: @escaping (Result<Bool, Error>) -> Void) {
    Task { @MainActor in
      do {
        try await AuthorizationCenter.shared.requestAuthorization(for: .individual)
        let authorized = AuthorizationCenter.shared.authorizationStatus == .approved
        completion(.success(authorized))
      } catch {
        completion(.failure(error))
      }
    }
  }

  func isAuthorized() -> Bool {
    return AuthorizationCenter.shared.authorizationStatus == .approved
  }

  func authorizationStatusString() -> String {
    switch AuthorizationCenter.shared.authorizationStatus {
    case .approved:
      return "authorized"
    case .denied:
      return "denied"
    case .notDetermined:
      return "not_determined"
    @unknown default:
      return "not_determined"
    }
  }

  func grantScreenTime(
    childUserId: String,
    durationMinutes: Int,
    allowedBundleIds: [String]?,
    allowedCategories: [String]?
  ) throws -> ScreenTimeGrantRecord {
    guard !childUserId.isEmpty else {
      throw ScreenTimeError.missingChildUserId
    }
    guard durationMinutes > 0 else {
      throw ScreenTimeError.invalidDuration
    }

    let now = Date()
    let expiresAt = now.addingTimeInterval(TimeInterval(durationMinutes * 60))
    let grant = ScreenTimeGrantRecord(
      grantId: UUID().uuidString,
      childUserId: childUserId,
      familyId: "",
      grantedMinutes: durationMinutes,
      grantedAt: now,
      expiresAt: expiresAt,
      status: "active",
      revokedAt: nil,
      allowedBundleIds: allowedBundleIds ?? [],
      allowedCategories: allowedCategories ?? [],
      grantedByUserId: "local",
      sourceTaskId: nil,
      usedMinutesOverride: nil
    )

    queue.sync {
      var grants = loadGrants()
      grants.removeAll { $0.childUserId == childUserId }
      grants.append(grant)
      saveGrants(grants)
    }

    applyGrantSettings()
    scheduleExpiryTimer()
    onUpdate?(grant)
    return grant
  }

  func revokeScreenTime(childUserId: String) throws -> ScreenTimeGrantRecord? {
    guard !childUserId.isEmpty else {
      throw ScreenTimeError.missingChildUserId
    }

    var revokedGrant: ScreenTimeGrantRecord?
    queue.sync {
      var grants = loadGrants()
      let now = Date()
      grants = grants.map { grant in
        guard grant.childUserId == childUserId else { return grant }
        var updated = grant
        updated.status = "revoked"
        updated.revokedAt = now
        updated.usedMinutesOverride = usedMinutes(for: updated, at: now)
        revokedGrant = updated
        return updated
      }
      grants.removeAll { $0.childUserId == childUserId }
      saveGrants(grants)
    }

    applyDefaultShielding()
    scheduleExpiryTimer()
    if let revokedGrant = revokedGrant {
      onUpdate?(revokedGrant)
    }
    return revokedGrant
  }

  func getScreenTimeStatus(childUserId: String?) -> (remainingMinutes: Int, activeGrants: [ScreenTimeGrantRecord]) {
    pruneExpiredGrants()
    let grants = queue.sync { loadGrants() }
    let activeGrants = grants.filter { grant in
      guard grant.status == "active" else { return false }
      if let childUserId = childUserId {
        return grant.childUserId == childUserId
      }
      return true
    }

    let now = Date()
    let remaining = activeGrants
      .map { remainingMinutes(for: $0, at: now) }
      .max() ?? 0

    return (remaining, activeGrants)
  }

  func syncGrantFromServer(grantData: [String: Any]) throws -> ScreenTimeGrantRecord {
    guard let grantId = grantData["grantId"] as? String,
          let childUserId = grantData["childUserId"] as? String,
          let familyId = grantData["familyId"] as? String else {
      throw ScreenTimeError.invalidPayload
    }

    let grantedMinutes = intValue(from: grantData["grantedMinutes"])
    guard grantedMinutes > 0 else {
      throw ScreenTimeError.invalidDuration
    }

    guard let grantedAt = dateValue(from: grantData["grantedAt"]),
          let expiresAt = dateValue(from: grantData["expiresAt"]) else {
      throw ScreenTimeError.invalidPayload
    }

    let status = (grantData["status"] as? String) ?? "active"
    let revokedAt = dateValue(from: grantData["revokedAt"])
    let allowedBundleIds = grantData["allowedAppBundleIds"] as? [String] ?? []
    let allowedCategories = grantData["allowedCategories"] as? [String] ?? []
    let grantedByUserId = grantData["grantedByUserId"] as? String ?? "server"
    let sourceTaskId = grantData["sourceTaskId"] as? String
    let usedMinutesOverride = intValue(from: grantData["usedMinutes"])

    var grant = ScreenTimeGrantRecord(
      grantId: grantId,
      childUserId: childUserId,
      familyId: familyId,
      grantedMinutes: grantedMinutes,
      grantedAt: grantedAt,
      expiresAt: expiresAt,
      status: status,
      revokedAt: revokedAt,
      allowedBundleIds: allowedBundleIds,
      allowedCategories: allowedCategories,
      grantedByUserId: grantedByUserId,
      sourceTaskId: sourceTaskId,
      usedMinutesOverride: usedMinutesOverride > 0 ? usedMinutesOverride : nil
    )

    let now = Date()
    if grant.status == "active", grant.expiresAt <= now {
      grant.status = "expired"
      grant.usedMinutesOverride = usedMinutes(for: grant, at: now)
    }

    queue.sync {
      var grants = loadGrants()
      grants.removeAll { $0.grantId == grantId }
      grants.append(grant)
      saveGrants(grants)
    }

    applyGrantSettings()
    scheduleExpiryTimer()
    onUpdate?(grant)
    return grant
  }

  func grantDictionary(_ grant: ScreenTimeGrantRecord) -> [String: Any] {
    let now = Date()
    let usedMinutes = usedMinutes(for: grant, at: now)
    let remainingMinutes = remainingMinutes(for: grant, at: now)
    var payload: [String: Any] = [
      "grantId": grant.grantId,
      "childUserId": grant.childUserId,
      "familyId": grant.familyId,
      "grantedMinutes": grant.grantedMinutes,
      "usedMinutes": usedMinutes,
      "remainingMinutes": remainingMinutes,
      "status": grant.status,
      "grantedAt": dateFormatter.string(from: grant.grantedAt),
      "expiresAt": dateFormatter.string(from: grant.expiresAt),
      "grantedByUserId": grant.grantedByUserId
    ]

    if let revokedAt = grant.revokedAt {
      payload["revokedAt"] = dateFormatter.string(from: revokedAt)
    }
    if let sourceTaskId = grant.sourceTaskId {
      payload["sourceTaskId"] = sourceTaskId
    }
    if !grant.allowedBundleIds.isEmpty {
      payload["allowedAppBundleIds"] = grant.allowedBundleIds
    }
    if !grant.allowedCategories.isEmpty {
      payload["allowedCategories"] = grant.allowedCategories
    }
    return payload
  }

  // MARK: - Private helpers

  private func usedMinutes(for grant: ScreenTimeGrantRecord, at date: Date) -> Int {
    if let override = grant.usedMinutesOverride {
      return min(grant.grantedMinutes, max(0, override))
    }
    let elapsed = max(0, date.timeIntervalSince(grant.grantedAt))
    return min(grant.grantedMinutes, Int(elapsed / 60))
  }

  private func remainingMinutes(for grant: ScreenTimeGrantRecord, at date: Date) -> Int {
    if grant.status != "active" {
      return 0
    }
    if date >= grant.expiresAt {
      return 0
    }
    return max(0, grant.grantedMinutes - usedMinutes(for: grant, at: date))
  }

  private func loadGrants() -> [ScreenTimeGrantRecord] {
    guard let data = UserDefaults.standard.data(forKey: defaultsKey) else {
      return []
    }
    let decoder = JSONDecoder()
    decoder.dateDecodingStrategy = .iso8601
    return (try? decoder.decode([ScreenTimeGrantRecord].self, from: data)) ?? []
  }

  private func saveGrants(_ grants: [ScreenTimeGrantRecord]) {
    let encoder = JSONEncoder()
    encoder.dateEncodingStrategy = .iso8601
    guard let data = try? encoder.encode(grants) else { return }
    UserDefaults.standard.set(data, forKey: defaultsKey)
  }

  private func pruneExpiredGrants() {
    queue.sync {
      let now = Date()
      var grants = loadGrants()
      var didExpire = false
      grants = grants.map { grant in
        guard grant.status == "active", grant.expiresAt <= now else { return grant }
        var updated = grant
        updated.status = "expired"
        updated.usedMinutesOverride = usedMinutes(for: updated, at: now)
        didExpire = true
        return updated
      }
      if didExpire {
        saveGrants(grants)
      }

      let stillActive = grants.filter { $0.status == "active" }
      if stillActive.isEmpty {
        applyDefaultShielding()
      }
    }
  }

  private func applyGrantSettings() {
    let now = Date()
    let grants = queue.sync { loadGrants() }
    let activeGrants = grants.filter { $0.status == "active" && $0.expiresAt > now }

    guard !activeGrants.isEmpty else {
      applyDefaultShielding()
      return
    }

    var allowAll = false
    var allowedBundleIds = Set<String>()
    var allowedCategories = Set<String>()

    for grant in activeGrants {
      if grant.allowedCategories.contains("all") {
        allowAll = true
        break
      }
      if grant.allowedBundleIds.isEmpty, grant.allowedCategories.isEmpty {
        allowAll = true
        break
      }
      allowedBundleIds.formUnion(grant.allowedBundleIds)
      allowedCategories.formUnion(grant.allowedCategories)
    }

    if allowAll {
      settingsStore.clearAllSettings()
      return
    }

    let applicationTokens = Set(allowedBundleIds.compactMap { bundleId in
      ApplicationToken(bundleIdentifier: bundleId)
    })

    let categoryTokens = Set(allowedCategories.compactMap { categoryId in
      categoryToken(from: categoryId)
    })

    settingsStore.clearAllSettings()

    if !applicationTokens.isEmpty {
      settingsStore.shield.applications = .all(except: applicationTokens)
      settingsStore.shield.applicationCategories = .none
    } else if !categoryTokens.isEmpty {
      settingsStore.shield.applications = .none
      settingsStore.shield.applicationCategories = .all(except: categoryTokens)
    } else {
      settingsStore.clearAllSettings()
    }
  }

  private func applyDefaultShielding() {
    settingsStore.clearAllSettings()
    settingsStore.shield.applications = .all
  }

  private func scheduleExpiryTimer() {
    timerQueue.sync {
      expiryTimer?.cancel()
      expiryTimer = nil
      let grants = queue.sync { loadGrants() }
      let activeGrants = grants.filter { $0.status == "active" }
      guard let nextExpiry = activeGrants.map(\.expiresAt).min() else { return }
      let interval = max(1, nextExpiry.timeIntervalSinceNow)
      let timer = DispatchSource.makeTimerSource(queue: timerQueue)
      timer.schedule(deadline: .now() + interval, leeway: .seconds(1))
      timer.setEventHandler { [weak self] in
        self?.expireGrantsIfNeeded()
      }
      expiryTimer = timer
      timer.resume()
    }
  }

  private func expireGrantsIfNeeded() {
    let now = Date()
    var expiredGrants: [ScreenTimeGrantRecord] = []
    queue.sync {
      var grants = loadGrants()
      grants = grants.map { grant in
        guard grant.status == "active", grant.expiresAt <= now else { return grant }
        var updated = grant
        updated.status = "expired"
        updated.usedMinutesOverride = usedMinutes(for: updated, at: now)
        expiredGrants.append(updated)
        return updated
      }
      if !expiredGrants.isEmpty {
        saveGrants(grants)
      }
    }

    if !expiredGrants.isEmpty {
      applyGrantSettings()
      expiredGrants.forEach { onUpdate?($0) }
    }
    scheduleExpiryTimer()
  }

  private func categoryToken(from identifier: String) -> ActivityCategoryToken? {
    switch identifier {
    case "games":
      return ActivityCategoryToken(.games)
    case "social":
      return ActivityCategoryToken(.socialNetworking)
    case "entertainment":
      return ActivityCategoryToken(.entertainment)
    case "education":
      return ActivityCategoryToken(.education)
    case "productivity":
      return ActivityCategoryToken(.productivity)
    default:
      return nil
    }
  }

  private func dateValue(from rawValue: Any?) -> Date? {
    if let date = rawValue as? Date {
      return date
    }
    if let string = rawValue as? String {
      return dateFormatter.date(from: string)
    }
    return nil
  }

  private func intValue(from rawValue: Any?) -> Int {
    if let value = rawValue as? NSNumber {
      return value.intValue
    }
    if let value = rawValue as? Int {
      return value
    }
    if let string = rawValue as? String, let parsed = Int(string) {
      return parsed
    }
    return 0
  }
}
